### PR TITLE
Adxrs290 Buffered channel support.

### PR DESCRIPTION
**Pull request by kister-jimenez**
To make this tinyiio demo as close as possible to the linux version, buffered channels has been added. The following are added:
1. Added buffered channel reading in the iio\iio_adxrs290 and the project.
2. Added the sync pin support to the driver to be able to check if the device has data available. 
3. Added a data_ready function from the driver and use this when getting data for the read_dev. This enables the read_dev to get the data only if it is available. Since TRIGGER is not yet supported by tinyiio, this acts as the throttling mechanism to ensure a minimum data reading interval.

This is how it will look like in IIO-OSC capture window (Auto scale disabled: Ymax: 30000, Ymin:-30000):
image

### CODE REVIEW

File: iio/iio_adxrs290/iio_adxrs290.c [hunk 1]
```diff
@@ -178,3 +179,61 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
...
+	uint32_t		k;
+	int16_t			data;
+
+	ch_idx = -1;
+	k = 0;
+	for (i = 0; i < nb_samples; i++)
+		while (get_next_ch_idx(adxrs290_ch_mask, ch_idx, &ch_idx)) {
+			adxrs290_get_rate_data(dev, ch_idx, &data);
+			if (ch_idx == ADXRS290_CHANNEL_TEMP)
+				data = (data << 4) >> 4;
```
Comment (mchindri) [inline iio/iio_adxrs290/iio_adxrs290.c:243]: nitpick: data &= 0xFFF0

File: drivers/gyro/adxrs290/adxrs290.c [hunk 2]
```diff
@@ -292,6 +292,13 @@ int32_t adxrs290_init(struct adxrs290_dev **device,
 	if (IS_ERR_VALUE(ret))
 		goto error_spi;
 
+	// Set GPIO sync pin. Set adxrs290 to output on sync pin.
+	gpio_get(&dev->gpio_sync, &init_param->gpio_sync);
```
Comment (mchindri) [inline drivers/gyro/adxrs290/adxrs290.c:319]: gpio_remove is needed in adxr290_remove

Comment (mchindri) [reply drivers/gyro/adxrs290/adxrs290.c:319]: Also check return values

File: drivers/gyro/adxrs290/adxrs290.c [hunk 3]
```diff
@@ -292,6 +292,13 @@ int32_t adxrs290_init(struct adxrs290_dev **device,
...
+	gpio_direction_input(dev->gpio_sync);
+	ret |= adxrs290_reg_write(dev, ADXRS290_REG_DATA_READY, ADXRS290_DATA_RDY_OUT);
```
Comment (mchindri) [inline drivers/gyro/adxrs290/adxrs290.c:321]: is this related to this commit?

File: projects/adxrs290-pmdz/src/main.c [hunk 4]
```diff
@@ -156,6 +159,12 @@ int main(void)
 		.spi_channel = SPI1
 	};
 
+	/* Aducm platform dependent initialization for GPIO. */
+	gpio_sync_init_param = (struct gpio_init_param) {
+		.number = 0x10,
```
Comment (mchindri) [inline projects/adxrs290-pmdz/src/main.c:164]: Use better a define for this value

File: iio/iio_adxrs290/iio_adxrs290.c (hunk 1 above)
Comment (kister-jimenez) [reply iio/iio_adxrs290/iio_adxrs290.c:243]: This is for sign-extension of temp value. The << will cause it to lose the 4 MSB. In the >> it go back to original form but MSB is now sign extended.

File: drivers/gyro/adxrs290/adxrs290.c (hunk 3 above)
Comment (kister-jimenez) [reply drivers/gyro/adxrs290/adxrs290.c:321]: Yes. The GPIO needs to be initialized as input and the device should be configured to output the DATA_READY on the sync pin. This DATA ready is needed so that it can be used in the filling of buffers in the iio_demo in the next commits.

File: iio/iio_adxrs290/iio_adxrs290.c (hunk 1 above)
Comment (mchindri) [reply iio/iio_adxrs290/iio_adxrs290.c:243]: sorry: I ment data &= 0xFFF

**Comment by kister-jimenez** [comment]
V2:
- define GPIO_SYNC_PIN_NUM 0x10 for the aducm3029 sync pin.
- Moved the gpio initialization in the main outside the ifdef ADUCM_PLATFORM since it's generic now.
- Checked returns of gpio routines and added gpio_remove() in the adxrs290_remove() and a goto label for error_gpio just before the error_spi.

File: iio/iio_adxrs290/iio_adxrs290.c [hunk 5]
```diff
@@ -178,3 +179,67 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
...
+			buff[k++] = data;
+		}
+	}
+
+	return nb_samples;
+}
```
Comment (amiclaus) [inline iio/iio_adxrs290/iio_adxrs290.c:245]: nitpick: new line at end of file

File: iio/iio_adxrs290/iio_adxrs290.c [hunk 6]
```diff
@@ -178,3 +179,57 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
...
+	while (!(ch_mask & 1)) {
+		last_idx++;
+		ch_mask >>= 1;
+	}
+	*new_idx = last_idx;
+
+	return 1;
+}
+
+int32_t	adxrs290_read_samples(void *device, uint16_t *buff,
```
Comment (adrimbarean) [inline iio/iio_adxrs290/iio_adxrs290.c:215]: nitpick: tab after int32_t

File: iio/iio_adxrs290/iio_adxrs290.c (hunk 5 above)
Comment (kister-jimenez) [reply iio/iio_adxrs290/iio_adxrs290.c]: Done.

File: iio/iio_adxrs290/iio_adxrs290.c [hunk 7]
```diff
@@ -49,6 +49,7 @@
 #define MAX_REG_ADDR		0x12
 
 volatile uint8_t current_direct_reg = 0;
+volatile uint8_t adxrs290_ch_mask = 0;
```
[the remark this answers is not in the document]
Comment (kister-jimenez) [reply iio/iio_adxrs290/iio_adxrs290.c]: Removed the remnants of the register routines.
Removed global variables and moved it to driver as a field of adxrs290. Added a routine to read all data in one SPI transaction.

File: iio/iio_adxrs290/iio_adxrs290.h [hunk 8]
```diff
@@ -133,10 +152,9 @@ struct iio_device adxrs290_iio_descriptor = {
 	.attributes = NULL,
 	.debug_attributes = NULL,
 	.buffer_attributes = NULL,
-	.transfer_dev_to_mem = NULL,
-	.transfer_mem_to_dev = NULL,
-	.read_data = NULL,
-	.write_data = NULL,
+	.prepare_transfer = adxrs290_update_active_channels,
+	.end_transfer = adxrs290_close_channels,
```
[the remark this answers is not in the document]
Comment (kister-jimenez) [reply iio/iio_adxrs290/iio_adxrs290.h]: Done.

File: iio/iio_adxrs290/iio_adxrs290.c (hunk 6 above)
Comment (kister-jimenez) [reply iio/iio_adxrs290/iio_adxrs290.c]: Done.

**Comment by kister-jimenez** [comment]
V3:
- Removed the global variables and other unused variables and prototypes in the iio_demo of adxrs290.
- Added the routine that can get all data in one SPI transaction to ensure that no data has changed  between multiple byte reads. [Datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/ADXRS290.pdf) section RATE OUTPUT DATA.

File: iio/iio_adxrs290/iio_adxrs290.c [hunk 9]
```diff
@@ -173,3 +173,30 @@ ssize_t set_adxrs290_iio_ch_lpf(void *device, char *buf, size_t len,
...
+{
+	struct adxrs290_dev	*dev = device;
+	uint32_t		i;
+	uint32_t		offset;
+	int16_t			data[ADXRS290_CHANNEL_COUNT];
+	uint8_t			ch_cnt;
+
+	offset = 0;
+	for (i = 0; i < nb_samples; i++) {
+		adxrs290_get_burst_data(dev, data, &ch_cnt);
```
Comment (mchindri) [inline iio/iio_adxrs290/iio_adxrs290.c:205]:
```suggestion
		adxrs290_get_burst_data(dev, data + offset, &ch_cnt);
```
And you will not need memcpy

File: drivers/gyro/adxrs290/adxrs290.c [hunk 10]
```diff
@@ -346,13 +369,30 @@ int32_t adxrs290_init(struct adxrs290_dev **device,
 	if (IS_ERR_VALUE(ret))
 		goto error_spi;
 
+	// Set GPIO sync pin.
+	ret |= gpio_get(&dev->gpio_sync, &init_param->gpio_sync);
```
Comment (mchindri) [inline drivers/gyro/adxrs290/adxrs290.c:373]: if you check every time I think is better to use = instead of |=

**Review by mchindri — APPROVED** [review]

**Review by amiclaus — APPROVED** [review]
